# 3. Write a Function for calculating the value of a European Option on Nifty using Monte Carlo Simulation for an expiry after a month. Compare it with the current market price and tell your decision, if it shall be bought or sold.

In [707]:
import datetime as dt
import numpy as np
import pandas as pd
from random import gauss
from math import sqrt, exp

In [708]:
def generate_asset_price(stockPrice, impliedVolatility, timeToMaturity, riskFreeRate):
    return stockPrice * exp((riskFreeRate - 0.5 * impliedVolatility ** 2) * timeToMaturity + impliedVolatility * sqrt(timeToMaturity) * gauss(0, 1)) 

In [709]:
def call_Payoff(stockPrice, strikePrice):
    return max(stockPrice - strikePrice, 0)

In [710]:
def put_Payoff(stockPrice, strikePrice):
    return max(strikePrice - stockPrice, 0)

In [711]:
def get_Values(stockPrice, strikePrice, impliedVolatility, timeToMaturity, riskFreeRate=0.1):
    return stockPrice, strikePrice, impliedVolatility, timeToMaturity, riskFreeRate

## Calculating Nifty Call Option Price on 28-June-2024
* Strike                   : 24,000 
* Current Underlying Price : 24,010.60
* Implied Volatility       : 11.64
* Option Expiry            : 25-July-2024
* Risk Free Rate           : 7.3%

In [712]:
numberOfSimulations = 10000

callPayOff = []

# Get stock values Time to maturity taken as 1-Month 
(stockPrice, strikePrice, impliedVolatility, timeToMaturity, riskFreeRate) = get_Values(stockPrice=24010.60, strikePrice=24000, impliedVolatility=0.1164, timeToMaturity=(21/252), riskFreeRate=0.073)

discountFactor = exp(-riskFreeRate + timeToMaturity)

for i in range(numberOfSimulations + 1):
    stockPrice_at_T = generate_asset_price(stockPrice, impliedVolatility, timeToMaturity, riskFreeRate)
    callPayOff.append(call_Payoff(stockPrice_at_T, strikePrice))

In [713]:
callOptionPrice = discountFactor * (sum(callPayOff) / numberOfSimulations)

print("Calculated Nifty Call Option Price : " + str(callOptionPrice))

Calculated Nifty Call Option Price : 414.19667648400747


### Q. Compare it with the current market price and tell your decision, if it shall be bought or sold. :
- Current Nifty call option Price as of 28-June-2024 is :  404.95
- Calculated Nifty Call Option Price : 414.19667648400747
- The calculated call option price is greater than current option  price ,Therefore the option is Undervalued.
- My Decision is to take a Long Position on the Call Option. By going Long on the Call Option we can book a profit.
